### PR TITLE
Add admin command functionality

### LINKED ISSUE #54: Add admin concept
For future commands which require administration privileges, I suggest that the application uses player objects instead of user_id's directly.

Player objects would have the is_admin attribute : 
```
class Player:
    """
    An object representation of a CTF player.
    """

    def __init__(self, user_id, is_admin):
        """
        user_id : The slack ID of a user
        is_admin : Is the user an administrator

        """
        self.user_id = user_id
        self.is_admin = is_admin
```

The help command would also show the appropriate menus based on the is_admin object.

### PR DESCRIPTION
**Pull request by Kileak**
Adds admin concept for bot commands.

Commands can now be flagged as admin commands and will only be visible for users in the `admin_users` group (defined in config.json).

New commands:
`add_admin [user_id]` - Add an user to the `admin_users` group (slack id is used not username)
`remove_admin [user_id]` - Removes an user from the `admin_users` group (slack id)
`show_admins` - Shows the user in `admin_users`

`archivectf` was changed to an admin command as an example, so it's only available to admin users.

On merging this change, the `admin_users` property should also be added to the current `config.json`, since we'll need initial admin users (Remember to add the slack id's (all uppercase) and not usernames).

Fixes #54

### CODE REVIEW

**Comment by Corb3nik** [comment]
Does adding/deleting an admin user support the `@username` notation?

I think the slack wrapper converts `@username` to the slack user ID. This should simplify admin management.

**Comment by Kileak** [comment]
Added parsing for @ notation

File: README.md [hunk 1]
```diff
@@ -30,13 +30,19 @@ Secondary features :
...
+!admin remove_admin <user_id>    (Remove an user from the admin user group)
+```
 
 ## Installation
 
 1. Copy `config.json.template` to `config.json`
 2. Fill the API token and bot name in the config.json file.
-3. `docker build -t ota-challenge-bot .`
-4. `docker run -it --rm --name live-ota-challenge-bot ota-challenge-bot`
+3. Add your user id (slack id, not the username) to `admin_users` group in `config.json`
```
Comment (Grazfather) [inline README.md]: How do I get this ID?

File: handlers/admin_handler.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            AdminHandler.send_message(slack_wrapper, channel_id, user_id, response)
+
+class AddAdminCommand(Command):
+    """Add an user to the admin user group."""
+    
+    def execute(self, slack_wrapper, args, channel_id, user_id):
+        """Execute the AddAdmin command."""
+        user = args[0].upper()
+
+        if (user.startswith("<@")) and (user.endswith(">")):
```
Comment (Grazfather) [inline handlers/admin_handler.py]: Don't see the outer parens

File: handlers/admin_handler.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            # TODO: Someone replace this with a nifty regex ;)
```
Comment (Grazfather) [inline handlers/admin_handler.py]: I'd remove: Your slicing is how I'd do it.

File: handlers/admin_handler.py (hunk 2 above)
Comment (Grazfather) [reply handlers/admin_handler.py]: We do this elsewhere. Maybe add a util to get a user object from a username or @username.

File: handlers/admin_handler.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                response = "User *{0}* is already in the admin group.".format(user_object['user']['name'])
+                AdminHandler.send_message(slack_wrapper, channel_id, user_id, response)
+        else:
+            response = "User *{0}* not found. You must provide the slack user id, not the username.".format(user)
+            AdminHandler.send_message(slack_wrapper, channel_id, user_id, response)
+
+        
+
+class RemoveAdminCommand(Command):
+    """Remove an user from the admin user group"""
```
Comment (Grazfather) [inline handlers/admin_handler.py]: Period missing.

Comment (Grazfather) [reply handlers/admin_handler.py]: Extra line.

File: handlers/admin_handler.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                response = "User *{0}* added to the admin group.".format(user_object['user']['name'])
+                AdminHandler.send_message(slack_wrapper, channel_id, user_id, response)
+            else:
+                response = "User *{0}* is already in the admin group.".format(user_object['user']['name'])
+                AdminHandler.send_message(slack_wrapper, channel_id, user_id, response)
+        else:
+            response = "User *{0}* not found. You must provide the slack user id, not the username.".format(user)
+            AdminHandler.send_message(slack_wrapper, channel_id, user_id, response)
+
+
```
Comment (Grazfather) [inline handlers/admin_handler.py]: Extra line.

File: handlers/admin_handler.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+class RemoveAdminCommand(Command):
+    """Remove an user from the admin user group"""
+    
+    def execute(self, slack_wrapper, args, channel_id, user_id):
+        """Execute the RemoveAdmin command."""
+        user = args[0].upper()
+
+        if (user.startswith("<@")) and (user.endswith(">")):
+            # TODO: Someone replace this with a nifty regex ;)
```
Comment (Grazfather) [inline handlers/admin_handler.py]: +1 for putting in util.

File: handlers/admin_handler.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+            user = user[2:-1]
+
+        admin_users = HandlerFactory.botserver.get_config_option("admin_users")
+
+        if admin_users and (user in admin_users):
+            admin_users.remove(user)
+            HandlerFactory.botserver.set_config_option("admin_users", admin_users)
+
+            response = "User *{0}* removed from the admin group.".format(user)
```
Comment (Grazfather) [inline handlers/admin_handler.py]: `{0}` -> `{}`

File: handlers/admin_handler.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    Commands :
+    # Show administrator users
+    !admin show_admins
+
+    # Add an user to the administrator group
+    !admin add_admin user_id
+
+    # Remove an user from the administrator group
+    !admin remove_admin user_id
+
```
Comment (Grazfather) [inline handlers/admin_handler.py]: extra line

File: handlers/admin_handler.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    """
+
+    # Specify if messages from syscall handler should be posted to channel (0)
+    # or per dm (1)
+    MSG_MODE = 0
+
+    def send_message(slack_wrapper, channel_id, user_id, msg):
```
Comment (Grazfather) [inline handlers/admin_handler.py]: No self?

File: handlers/admin_handler.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        """Send message to user or channel, depending on configuration."""
+        dest_channel = channel_id if (
```
Comment (Grazfather) [inline handlers/admin_handler.py]: Don't need parens.

File: handlers/base_handler.py [hunk 11]
```diff
@@ -6,8 +6,16 @@
...
 
-    def can_handle(self, command):
-        return command in self.commands
+    def can_handle(self, command, user_is_admin):
+        cmd_available = command in self.commands
+
+        # Check if this is an admin command and hide it for normal users
+        if cmd_available:
+            cmd_desc = self.commands[command]
+            if (not cmd_desc.is_admin_cmd) or user_is_admin:
```
Comment (Grazfather) [inline handlers/base_handler.py]: parens not needed.

File: handlers/base_handler.py [hunk 12]
```diff
@@ -19,42 +27,46 @@ def parse_command_usage(self, command, descriptor):
...
+    def get_usage(self, user_is_admin):
         """Return the usage of a handler."""
         msg = "```"
 
         for command in self.commands:
             descriptor = self.commands[command]
-            usage = self.parse_command_usage(command, descriptor)
-            msg += "!{} {}\n".format(self.handler_name, usage)
+
+            if (not descriptor.is_admin_cmd) or (user_is_admin):
```
Comment (Grazfather) [inline handlers/base_handler.py]: parens.

File: handlers/base_handler.py [hunk 13]
```diff
@@ -19,42 +27,46 @@ def parse_command_usage(self, command, descriptor):
...
         return msg
 
-    def process(self, slack_wrapper, command, args, channel, user):
+    def process(self, slack_wrapper, command, args, channel, user, user_is_admin):
         """Check if enough arguments were passed for this command."""
         cmd_descriptor = self.commands[command]
 
         if cmd_descriptor:
-            if (cmd_descriptor.arguments) and (len(args) < len(cmd_descriptor.arguments)):
+            if (cmd_descriptor.arguments) and (len(args) < len(cmd_descriptor.arguments)):
```
Comment (Grazfather) [inline handlers/base_handler.py]: Trailing whitespace, extra parens.

File: handlers/base_handler.py [hunk 14]
```diff
@@ -19,42 +27,46 @@ def parse_command_usage(self, command, descriptor):
...
         """Check if enough arguments were passed for this command."""
         cmd_descriptor = self.commands[command]
 
         if cmd_descriptor:
-            if (cmd_descriptor.arguments) and (len(args) < len(cmd_descriptor.arguments)):
+            if (cmd_descriptor.arguments) and (len(args) < len(cmd_descriptor.arguments)):                
                 raise InvalidCommand(
                     self.command_usage(command, cmd_descriptor))
-            else:
+            else:
```
Comment (Grazfather) [inline handlers/base_handler.py]: whitespace

File: handlers/handler_factory.py [hunk 15]
```diff
@@ -45,6 +47,9 @@ def process(slack_wrapper, botserver, message, channel_id, user_id):
             processed = False
             usage_msg = ""
 
+            admin_users = botserver.get_config_option("admin_users")
+            user_is_admin = (admin_users) and (user_id in admin_users)
```
Comment (Grazfather) [inline handlers/handler_factory.py]: parens

File: server/botserver.py [hunk 16]
```diff
@@ -101,7 +102,8 @@ def run(self):
         while self.running:
             try:
                 self.load_config()
-                self.slack_wrapper = SlackWrapper(self.get_config_option('api_key'))
+                self.slack_wrapper = SlackWrapper(
+                    self.get_config_option('api_key'))
```
Comment (Grazfather) [inline server/botserver.py]: Use `"` like everywhere else.

File: config.json.template [hunk 17]
```diff
@@ -2,4 +2,5 @@
   "api_key" : "",
   "bot_name" : "",
   "send_help_as_dm" : "1"
+  "admin_users" : []
```
Comment (Grazfather) [inline config.json.template]: Consider using a set

File: handlers/admin_handler.py (hunk 4 above)
Comment (Grazfather) [reply handlers/admin_handler.py]: typo "an user" -> "a user"

File: handlers/base_handler.py (hunk 11 above)
Comment (Kileak) [reply handlers/base_handler.py]: Won't fix.

File: server/botserver.py (hunk 16 above)
Comment (Kileak) [reply server/botserver.py]: This is just a reformat?

Comment (Grazfather) [reply server/botserver.py]: Yes, but might as well correct if you're editing the line.

File: util/util.py [hunk 18]
```diff
@@ -179,3 +180,20 @@ def remove_ctf_by_channel_id(database, ctf_channel_id):
     ctf = ctfs[ctf_channel_id]
     ctfs.pop(ctf_channel_id)
     pickle.dump(ctfs, open(database, "wb"))
+
+
+def parse_user_id(user_id):
+    """
+    Parses a user_id, removing possible @-notation and make sure it's uppercase
```
Comment (Grazfather) [inline util/util.py]: You're gonna kill me for all these nitpicks, but: 'Parse' vs. 'Parses'

File: util/util.py [hunk 19]
```diff
@@ -179,3 +180,20 @@ def remove_ctf_by_channel_id(database, ctf_channel_id):
...
+    """
+    if user_id.startswith("<@") and user_id.endswith(">"):
+        return user_id[2:-1].upper()
+
+    return user_id.upper()
+
+
+def resolve_user_by_user_id(slack_wrapper, user_id):
+    """
+    Resolves an user id to an user object.
```
Comment (Grazfather) [inline util/util.py]: 'Resolve' vs. 'Resolves'

File: util/util.py [hunk 20]
```diff
@@ -167,7 +167,8 @@ def remove_challenge_by_channel_id(database, challenge_channel_id, ctf_channel_i
     """
     ctfs = pickle.load(open(database, "rb"))
     ctf = ctfs[ctf_channel_id]
-    ctf.challenges = list(filter(lambda challenge: challenge.channel_id != challenge_channel_id, ctf.challenges))
+    ctf.challenges = list(filter(
```
Comment (Grazfather) [inline util/util.py]: Same as: `[challenge for challenge in ctf.challenges if challenge.channel_id != challenge_channel_id]`.

You can keep as is, I find the list comprehensions to be easier to read.

Comment (Grazfather) [reply util/util.py]: This is just a reformat, so let's keep it for now.

File: util/util.py (hunk 18 above)
Comment (Kileak) [reply util/util.py]: Nah, it's right, just not used to python stlyle / docstrings. But I would point those things out, too in a code review, if they violate a predefined style guide.

**Comment by Kileak** [comment]
Not sure enough on those lambda expressions, so I leave them as they are. 

Util has to be revisited anyways, since it contains several errors (a player object was introduced, but seems only halfway implemented), but should be done in another pr.

**Comment by Grazfather** [comment]
LGTM

**Pull request merged**